atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStoreTest.java

```diff
@@ -120,66 +138,217 @@ public void returnsTimestampAssociatedWithGreatestPunchedTimeEvenIfItIsNotGreate
+        verify(kvs, atMost(maxReads)).getRange(eq(AtlasDbConstants.PUNCH_TABLE), any(RangeRequest.class), anyLong());
+    }
+
+    @Test
+    public void getMillisForTimestampWithinBoundsNoExactMatch() {
+        KeyValueService kvs = Mockito.spy(new InMemoryKeyValueService(false));
+        PuncherStore puncherStore = KeyValueServicePuncherStore.create(kvs);
+
+        int generousReadLimit = LongMath.log2(MAX_OFFSET_FOR_BOUNDS_TESTS, RoundingMode.CEILING) * 5;
+        long timestampBound = 1_000_000;
+        long queryTimestamp = timestampBound * 2 / 3;
+
+        Multimap<Long, Long> tsToMillis = generateTsToMillisMap(timestampBound, false);
+        tsToMillis.forEach(puncherStore::put);
+
+        Collection<Long> millisCandidates = tsToMillis.keySet().stream()
+                .filter(ts -> ts <= queryTimestamp)
+                .max(Long::compareTo)
+                .map(tsToMillis::get)
+                .get();
+        Optional<MillisAndMaybeTimestamp> actual = KeyValueServicePuncherStore.getMillisForTimestampWithinBounds(
+                kvs, queryTimestamp, null, MAX_OFFSET_FOR_BOUNDS_TESTS);
+
+        assertThat(actual).isPresent();
+        assertThat(millisCandidates).contains(actual.get().millis());
+        assertThat(actual.get().timestamp()).isPresent();
+        assertThat(actual.get().timestamp().get()).isLessThan(queryTimestamp);
+
+        verify(kvs, atMost(generousReadLimit))
+                .getRange(eq(AtlasDbConstants.PUNCH_TABLE), any(RangeRequest.class), anyLong());
+
+        ArgumentCaptor<RangeRequest> rangeRequest = ArgumentCaptor.forClass(RangeRequest.class);
+        verify(kvs).getRange(eq(AtlasDbConstants.PUNCH_TABLE), rangeRequest.capture(), eq(queryTimestamp + 1));
+        assertThat(getRangeSize(rangeRequest.getValue()))
+                .isBetween(1L, KeyValueServicePuncherStore.MAX_RANGE_SCAN_SIZE);
+    }
+
+    @Test
+    public void getMillisForTimestampWithinBoundsWithExactMatch() {
+        KeyValueService kvs = Mockito.spy(new InMemoryKeyValueService(false));
+        PuncherStore puncherStore = KeyValueServicePuncherStore.create(kvs);
+
+        int generousReadLimit = LongMath.log2(MAX_OFFSET_FOR_BOUNDS_TESTS, RoundingMode.CEILING) * 5;
+        long timestampBound = 1_000_000;
+        long queryTimestamp = timestampBound * 2 / 3;
+
+        Multimap<Long, Long> tsToMillis = generateTsToMillisMap(timestampBound, true);
+        tsToMillis.forEach(puncherStore::put);
+
+        Collection<Long> millisCandidates = tsToMillis.get(queryTimestamp);
+
+        Optional<MillisAndMaybeTimestamp> actual = KeyValueServicePuncherStore.getMillisForTimestampWithinBounds(
+                kvs, queryTimestamp, null, MAX_OFFSET_FOR_BOUNDS_TESTS);
+
+        assertThat(actual).isPresent();
+        assertThat(millisCandidates).contains(actual.get().millis());
+        assertThat(actual.get().timestamp()).hasValue(queryTimestamp);
```

Comment:
what happens if we have more than one match for the timestamp?

Reply:
fixed